### ISSUE TITLE
Answer: How to use TanhGrad (with test sample)

### ISSUE THREAD

**Issue by EtcEc** [migueldeicaza/TensorFlowSharp#131]
Reproduce error like this:

TFGraph graph = new TFGraph();
TFSession session = new TFSession(graph);

TFOutput x = graph.Placeholder(TFDataType.Double, new TFShape(1));
TFOutput y = graph.Tanh(x);
TFOutput dy = graph.TanhGrad(x, y);

TFTensor[] result = session.Run(new[] {x}, new[] {new TFTensor(0.7)}, new[] {dy});

Console.WriteLine(((double) result[0].GetValue()).ToString());

This produces the result: 0,308227566329753
but should produce: 0.63474
https://www.wolframalpha.com/input/?i=d%2Fdx+tanh(x)+at+0.7

**Comment by Dorokhov** [comment]
Hi @EtcEc, 
I've added an example for TanhGrad() API in PR #132 based on your code.

In general, TanhGrad() API requires two parameters: 'y' and 'dy' where 
y = tanh(x), and 'dy' is the corresponding input gradient: TFOutput grad = graph.TanhGrad (y, dy);
Considering that, your could should look like this

TFOutput x = graph.Placeholder(TFDataType.Double, TFShape.Scalar);
TFOutput y = graph.Tanh(x);
**TFOutput dy = graph.Const (new TFTensor (new [] { 1.0 }));
TFOutput grad = graph.TanhGrad(y, dy);**

Here's visualization of 'grad' calculation
https://www.wolframalpha.com/input/?i=1*+(1-+tanh(x)+*+tanh(x))+at+0.7
As you can see it produces exactly the same result you expected.

**Comment by EtcEc** [comment]
Thank You for the helpful answer. This does indeed produce the correct result!

However, i suggest either of two things:

1) updating to appropriate naming of the inputs and the doc-string. Currently You have

TanhGrad(TFOutput x, TFOutput y) // Computes the gradient for the tanh of `x` wrt its input.

Instead i suggest

TanhGrad(TFOutput y, TFOutput dy) // Computes the gradient for y = tanh(x) w.r.t. x, backpropagating the delta dy.

I suggest this description since, from what you referred to on wolfram, this method not only calculates the gradient, but also multiplies with dy.

**Comment by Dorokhov** [comment]
Hi @EtcEc,

You highlighted a good point. The problem is that all tensorflow operation signatures and documentation defined in native library and source codes lie in original tensorflow framework repository. So we need to address the problem there.

I created [issue](https://github.com/tensorflow/tensorflow/issues/12902) asking to rename parameters in native API, will see what they answer.

Thanks.
